experimental/cuda2/pending_queue_actions.c

```diff
@@ -197,10 +294,23 @@ void iree_hal_cuda2_pending_queue_actions_destroy(
   iree_hal_cuda2_pending_queue_actions_t* actions =
       iree_hal_cuda2_pending_queue_actions_cast(base_actions);
   iree_allocator_t host_allocator = actions->host_allocator;
+  iree_hal_cuda2_working_area_t* working_area = &actions->working_area;
   IREE_TRACE_ZONE_BEGIN(z0);
 
   IREE_ASSERT(iree_hal_cuda2_queue_action_list_is_empty(&actions->action_list));
 
+  iree_atomic_int32_t* exiting_state = &working_area->exiting_state;
+  iree_atomic_store_int32(exiting_state,
+                          IREE_HAL_CUDA_WORKER_STATE_EXIT_REQUESTED,
+                          iree_memory_order_relaxed);
+  iree_notification_post(&working_area->notification, IREE_ALL_WAITERS);
+  while (iree_atomic_load_int32(exiting_state, iree_memory_order_relaxed) !=
+         IREE_HAL_CUDA_WORKER_STATE_EXIT_COMMITTED) {
+    // Busy wait until the worker thread exits.
```

Comment:
you need to wait in here as it's possible to starve the thread
instead, you can wait on the notification in a loop and have the thread post the notification just before exiting

you basically are just waiting until a point in time when you know the worker won't access any of the stuff you're about to free below

see for example iree_task_poller_await_exit / iree_task_poller_main

Reply:
Ah, nice!! Actually you have iree_task_poller_await_exit / iree_task_poller_main already doing similar things.. I should've checked that before writing this up.. Good that I didn't come up something totally different. ;-P